lib/TransactionRunner.js

```diff
@@ -200,137 +183,14 @@ class TransactionRunner {
-
-var _func = ${hookCode};
-_func(_data);
-
-// setup the return object
-var output = {};
-output["data"] = _data;
-output["stash"] = stash;
-output["logs"] = _logs;
-output["now"] = Date.now();
-output;\
-`;
-  }
-
-  runSandboxedHookFromString(hookString, data, callback) {
-    const wrappedCode = this.sandboxedWrappedCode(hookString);
-
-    const sandbox = new Pitboss(wrappedCode, { timeout: 500 });
-    return sandbox.run({
-      context: {
-        _data: data,
-        _logs: [],
-        stash: this.hookStash,
-      },
-      libraries: {
-        _log: sandboxedLogLibraryPath,
-      },
-    },
-    (err, result = {}) => {
-      sandbox.kill();
-      this.sandboxedHookResultsHandler(err, data, result, callback);
-    });
-  }
-
-  // Will be used runHook instead in next major release, see deprecation warning
-  runLegacyHook(hook, data, callback) {
-    // Not sandboxed mode - hook is a function
-    if (typeof hook === 'function') {
-      if (hook.length === 1) {
-        // Sync api
-        logger.warn(`\
-DEPRECATION WARNING!
-
-You are using only one argument for the \`beforeAll\` or \`afterAll\` hook function.
-One argument hook functions will be treated as synchronous in the near future.
-To keep the async behaviour, just define hook function with two parameters.
-
-Interface of the hooks functions will be unified soon across all hook functions:
-
- - \`beforeAll\` and \`afterAll\` hooks will support sync API depending on number of arguments
- - Signatures of callbacks of all hooks will be the same
- - First passed argument will be a \`transactions\` object
- - Second passed argument will be a optional callback function for async
- - \`transactions\` object in \`hooks\` module object will be removed
- - Manipulation with transaction data will have to be performed on the first function argument\
-`);
-
-        // DEPRECATION WARNING
-        // This will not be supported in future hook function will be called with
-        // data synchronously and callback will be called immediatelly and not
-        // passed as a second argument
-        hook(callback);
-      } else if (hook.length === 2) {
-        // Async api
-        hook(data, () => callback());
-      }
-    }
-
-    // Sandboxed mode - hook is a string - only sync API
-    if (typeof hook === 'string') {
-      this.runSandboxedHookFromString(hook, data, callback);
-    }
-  }
-
   runHook(hook, data, callback) {
-    // Not sandboxed mode - hook is a function
-    if (typeof hook === 'function') {
-      if (hook.length === 1) {
-        // Sync api
-        hook(data);
-        callback();
-      } else if (hook.length === 2) {
-        // Async api
-        hook(data, () => callback());
-      }
-    }
-
-    // Sandboxed mode - hook is a string - only sync API
-    if (typeof hook === 'string') {
-      this.runSandboxedHookFromString(hook, data, callback);
+    if (hook.length === 1) {
+      // Sync api
+      hook(data);
+      callback();
```

Comment:
This reminds me how very simple it would be to support async/promise returning callbacks:

```js
const result = callback();

// Once Node >= 10 use https://nodejs.org/api/util.html#util_util_types_ispromise_value
// Or perhaps only support Promises if Node >= 8 or Node >= 10 to simplify code
if (/* node >= 8.2 where callbackify is existant? */ && result && result.then) {
  const {callbackify} = require('util');
  callbackify(result)(() => callback()); // shouldn't we support errors? I saw below async API doesn't?
} else {
  callback();
}
```

Reply:
My plans are to revamp the JS hooks in some future after this. As I commented above, I can even imagine someone contributing relevant changes to the JS hooks when this gets merged. My ideas around the JS hooks API would be:

- sync support (no callback parameter, no return value)
- async callback support (accepts callback)
- async promise support (returns promise)
- tested async/await support
- `transaction.fail = ...` gracefully fails the hook
- thrown error gracefully fails the hook (both sync and async)
- `callback(error)` gracefully fails the hook
- rejected promise fails the hook

Also thinking whether there's a way to make the hooks pure functions in the future, but I'd leave that aside for now as such change is a big one and sort of spans across the languages.